app/invocation/invocation_action_tree_node.tsx

```diff
@@ -20,7 +20,7 @@ type SymlinkNode = build.bazel.remote.execution.v2.SymlinkNode;
 
 export interface TreeNode {
   obj: FileNode | DirectoryNode | SymlinkNode;
-  type: "file" | "dir" | "symlink";
+  type: "file" | "dir" | "symlink" | "tree";
 }
```

Comment:
I missed this in the earlier PR, but this looks like an ideal use case for the [discriminated union](https://www.typescriptlang.org/docs/handbook/unions-and-intersections.html#discriminating-unions) pattern.

```typescript
export type TreeNode = {
  obj: FileNode;
  type: "file";
} | {
  obj: DirectoryNode;
  type: "directory";
} | {
  obj: SymlinkNode;
  type: "symlink";
} | {
  obj: DirectoryNode;
  type: "tree";
}
```

If you defined the type like that, then you could get some nicer type narrowing:

```typescript
if (node.type === "symlink") {
  // Within this `if` block, TS narrows the type of `node` to `{ obj: SymlinkNode; type: "symlink" }`

  doSomethingWith(obj.digest); // This doesn't compile, because symlink doesn't have a digest field
}
```

Reply:
Great suggestion! 🎉 

I changed it to union type.